packages/flutter/lib/src/material/text_selection.dart

```diff
@@ -20,64 +23,417 @@ const double _kHandleSize = 22.0;
+    if (_key.currentContext != null) {
+      final RenderBox renderBoxChild = _key.currentContext.findRenderObject() as RenderBox;
+      if (_maxWidth == null || renderBoxChild.size.width > _maxWidth) {
+        _maxWidth = renderBoxChild.size.width;
+      }
+    }
+
     final MaterialLocalizations localizations = MaterialLocalizations.of(context);
     final List<Widget> items = <Widget>[
-      if (handleCut != null) FlatButton(child: Text(localizations.cutButtonLabel), onPressed: handleCut),
-      if (handleCopy != null) FlatButton(child: Text(localizations.copyButtonLabel), onPressed: handleCopy),
-      if (handlePaste != null) FlatButton(child: Text(localizations.pasteButtonLabel), onPressed: handlePaste),
-      if (handleSelectAll != null) FlatButton(child: Text(localizations.selectAllButtonLabel), onPressed: handleSelectAll),
+      if (widget.handleCut != null)
+        _getItem(widget.handleCut, localizations.cutButtonLabel),
+      if (widget.handleCopy != null)
+        _getItem(widget.handleCopy, localizations.copyButtonLabel),
+      if (widget.handlePaste != null)
+        _getItem(widget.handlePaste, localizations.pasteButtonLabel),
+      if (widget.handleSelectAll != null)
+        _getItem(widget.handleSelectAll, localizations.selectAllButtonLabel),
     ];
 
     // If there is no option available, build an empty widget.
     if (items.isEmpty) {
       return Container(width: 0.0, height: 0.0);
     }
 
-    return Material(
-      elevation: 1.0,
-      child: Container(
-        height: _kToolbarHeight,
-        child: Row(mainAxisSize: MainAxisSize.min, children: items),
+    final Container child = Container(
+      key: _key,
+      height: _overflowOpen ? null : _kToolbarHeight,
+      child: Material(
+        elevation: 1.0,
+        child: _TextSelectionToolbarItems(
+          isAbove: widget.isAbove,
+          overflowOpen: _overflowOpen,
+          children: <Widget>[
+            // The navButton that shows and hides the overflow menu is the first
+            // child.
+            Material(
+              child: IconButton(
+                // TODO(justinmc): This should be an AnimatedIcon, but
+                // AnimatedIcons doesn't yet support arrow_back to more_vert.
+                // https://github.com/flutter/flutter/issues/51209
+                icon: Icon(_overflowOpen ? Icons.arrow_back : Icons.more_vert),
+                onPressed: () {
+                  setState(() {
+                    _overflowOpen = !_overflowOpen;
+                  });
+                },
+                tooltip: _overflowOpen ? 'Back' : 'More',
+              ),
+            ),
+            ...items,
+          ],
+        ),
+      ),
+    );
+
+    // When the menu shrinks as the overflow is opened, then it occupies the
+    // same size as it did with overflow closed and aligns itself to the right
+    // of its original size.
+    return _maxWidth == null ? child : SizedBox(
```

Comment:
I don't fully get the _maxWidth business. Isn't that value always (at least) one frame behind?